packages/w3p/src/provider.ts

```diff
@@ -0,0 +1,214 @@
+          value as (e: ProviderEventPayload) => void,
+        )
+      })
+    }
+
+    this.#selectedProvider = provider.name
+    await this.#proxy?.init()
+    return this
+  }
+
+  public async connect() {
+    if (!this.#proxy) throw new errors.ProviderNotInitializedError()
+
+    await this.#proxy.connect()
+  }
+
+  public async switchChain(chainId: ChainId) {
+    await this.#proxy?.switchChain?.(chainId)
+  }
+
+  public async addChain(chain: Chain) {
+    await this.#proxy?.addChain?.(chain)
+  }
+
+  public setChainDetails(chain: Chain) {
+    this.#chainDetails = chain
+  }
+
+  public async signAndSendTx(txRequestBody: TxRequestBody) {
+    if (this.#proxy?.signAndSendTx) {
+      return this.#proxy?.signAndSendTx?.(
+        txRequestBody,
+      ) as Promise<TransactionResponse>
+    }
+
+    throw new errors.ProviderMethodNotSupported()
+  }
+
+  public getHashFromTx(txResponse: TransactionResponse) {
+    return this.#proxy?.getHashFromTx?.(txResponse) ?? ''
+  }
+
+  public getTxUrl(chain: Chain, txHash: string) {
+    return this.#proxy?.getTxUrl?.(chain, txHash) ?? ''
+  }
+
+  public getAddressUrl(chain: Chain, address: string) {
+    return this.#proxy?.getAddressUrl?.(chain, address) ?? ''
+  }
+
+  public async signMessage(message: string) {
+    return this.#proxy?.signMessage?.(message) ?? ''
+  }
+
+  public onAccountChanged(cb: (e: ProviderEventPayload) => void): void {
+    this.#proxy?.onAccountChanged(cb)
+  }
+
+  public onChainChanged(cb: (e: ProviderEventPayload) => void): void {
+    this.#chainDetails = undefined
+
+    this.#proxy?.onChainChanged?.(cb)
+  }
+
+  public onConnect(cb: (e: ProviderEventPayload) => void): void {
+    this.#proxy?.onConnect(cb)
+  }
+
+  public onDisconnect(cb: (e: ProviderEventPayload) => void): void {
+    this.#proxy?.onDisconnect(cb)
+  }
+
+  public onInitiated(cb: (e: ProviderEventPayload) => void): void {
+    this.#proxy?.onInitiated(cb)
+  }
+
+  public clearHandlers(): void {
+    this.#proxy?.clearHandlers()
+  }
+
+  public onBeforeTxSent(cb: (e: ProviderEventPayload) => void) {
+    this.#proxy?.onBeforeTxSent(cb)
+  }
+
+  public onAfterTxSent(cb: (e: ProviderEventPayload) => void) {
+    this.#proxy?.onAfterTxSent(cb)
+  }
+
+  public onAfterTxConfirmed(cb: (e: ProviderEventPayload) => void) {
+    this.#proxy?.onAfterTxConfirmed(cb)
+  }
+
+  public async disconnect() {
+    if (this.#proxy?.disconnect) {
+      await this.#proxy?.disconnect()
+    }
+
+    throw new errors.ProviderMethodNotSupported()
+  }
+}
+
+/**
+ * @description Creates an instance of a wallet provider
+ *
+ * @example
+ * const provider = await createProvider(MetamaskProvider)
+ * // or
+ * const providerDetectorInstance = await new providerDetector().init()
+ * const metamaskProvider = await createProvider(MetamaskProvider, { providerDetectorInstance })
+ * const phantomProvider = await createProvider(PhantomProvider, { providerDetectorInstance })
```

Comment:
let's rename opt to just `providerDetector` and fix `await new providerDetector().init()` to `await new ProviderDetector().init()`